juju/client/connection.py

```diff
@@ -553,7 +645,23 @@ def connect_params(self):
     def _build_facades(self, facades):
         self.facades.clear()
         for facade in facades:
-            self.facades[facade['name']] = facade['versions'][-1]
+            name = facade['name']
+            # the following attempts to get the best facade version for the
+            # client. The client knows about the best facade versions it speaks,
+            # so in order to be compatible forwards and backwards we speak a
+            # common facade versions.
+            if name in client_facades:
+                try:
+                    known = client_facades[name]['versions']
+                    discovered = facade['versions']
+                    version = max(set(known).intersection(set(discovered)))
+                except ValueError:
+                    # this can occur if known is [1, 2] and discovered is [3, 4]
+                    # there is just know way to know how to communicate with the
+                    # facades we're trying to call.
+                    raise errors.JujuConnectionError('unknown common facade version')
```

Comment:
My concern with raising an exception here is that all of these facades are being looked up *before* we actually try to use them. It seems a shame to prevent being able to operate against a server *at all* if *one* facade is not supported. I would be happy with it if we gave this sort of error when a shim calls "GetFacade('name')", but not in an 'iterate everything first.
I think its probably better to 
a) log a warning?
b) just treat the facade as not present.

Reply:
log a warning, feels better.